lib/logstash/outputs/elasticsearch.rb

```diff
@@ -426,55 +539,51 @@ def generate_jks cert_path
     [jks.path, pwd]
   end
 
-  public
-  def receive(event)
-    return unless output?(event)
-
-    # Set the 'type' value for the index.
-    if @index_type
-      type = event.sprintf(@index_type)
-    else
-      type = event["type"] || "logs"
+  private
+  # in charge of submitting any actions in @retry_queue that need to be 
+  # retried
+  #
+  # This method is not called concurrently. It is only called by @retry_thread
+  # and once that thread is ended during the teardown process, a final call 
+  # to this method is done upon teardown in the main thread.
+  def retry_flush()
+    # it is only necessary to lock this mutex for the process of waiting for a 
+    # signal, and does not need to wrap the rest of the flush execution because 
+    # we ensure this method is to not be called concurrently.
+    if @retry_teardown_requested.false?
+      @retry_flush_mutex.synchronize { @retry_queue_needs_flushing.wait(@retry_flush_mutex) }
     end
 
-    index = event.sprintf(@index)
+    unless @retry_queue.empty?
+      buffer = @retry_queue.size.times.map do
+        next_action, next_doc, next_event = @retry_queue.pop
+        next_event['@metadata']['retry_count'] += 1
 
-    document_id = @document_id ? event.sprintf(@document_id) : nil
-    buffer_receive([event.sprintf(@action), { :_id => document_id, :_index => index, :_type => type }, event.to_hash])
-  end # def receive
+        if next_event['@metadata']['retry_count'] > @max_retries
+          @logger.error "too many attempts at sending event. dropping: #{next_event}"
```

Comment:
until we have the DLQ, wondering if we should log more upon dropping the event? 
```ruby
#{next_event.to_hash_with_metadata.inspect}
```
?? 

Any thoughts @jordansissel ?